utils/parametric-components.stanza

```diff
@@ -0,0 +1,537 @@
+    set-property(self, to-symbol(myk), myv)
+
+  ; possible properties to extract from params:
+  ; area
+  ; case
+  ; category
+  ; composition
+  ; dimensions.area
+  ; dimensions.x
+  ; dimensions.y
+  ; dimensions.z
+  ; manufacturer
+  ; metadata.datasheets
+  ; metadata.description
+  ; metadata.digi-key-part-number
+  ; metadata.features
+  ; metadata.image
+  ; metadata.number-of-terminations
+  ; metadata.packaging
+  ; metadata.series
+  ; metadata.supplier-device-package
+  ; minimum_quantity
+  ; mounting
+  ; mpn
+  ; price
+  ; rated-power
+  ; rated-temperature.max
+  ; rated-temperature.min
+  ; resistance
+  ; stock
+  ; tcr.neg
+  ; tcr.pos
+  ; tolerance
+  ; tolerance.max
+  ; tolerance.min
+  ; trust
+  ; type
+  ; update_date
+  ; vendor_part_numbers.<VENDOR_STRING>
+  ; x
+  ; y
+  ; z
+
+  ; re-enable when spice is available in jitx-client Docker image for CI
+  ; spice:
+  ;   "[R] <p[1]> <p[2]> <property(self.resistance)>"
+
+public defn parametric-resistor (res:Double|Toleranced) :
+  parametric-resistor(["resistance" => res])
+
+public defn parametric-resistor (res:Double|Toleranced, tol:Double, pkg:String) :
+  parametric-resistor(["resistance" => res "tolerance" => tol "rated-power" => 0.0625 "case" => pkg])
+
+public defn parametric-resistor (res:Double|Toleranced, tol:Double, pwr:Double) :
+  parametric-resistor(["resistance" => res "tolerance" => tol "rated-power" => pwr "case" => "0402"])
+
+public defn parametric-resistor (res:Double|Toleranced, pkg:String) :
+  parametric-resistor(["resistance" => res "tolerance" => 2.0 "rated-power" => 0.0625 "case" => pkg])
+
+public defn parametric-resistor (res:Double|Toleranced, tol:Double) -> Instantiable :
+  parametric-resistor(["resistance" => res "tolerance" => tol])
+
+
+; ========================================================
+; ========== parametric part helper functions ============
+; ========================================================
+defn c-r (v:Double e:Int) :
+  v * pow(10.0, to-double(e))
+
+
+  ; Returns a value within the range of v in the tolerance range specified by tol
+public defn between-std-val (v:Toleranced, tol:Double) -> Double :
+  val expon-min = floor(log10(min-value(v)))
+  val expon-max = floor(log10(max-value(v)))    
+  var match-exp = 0.0
+  var match-val = -1.0
+  var match-diff = 1.0e6
+  ; println("Need to find exp range (%_ %_) for range (%_)" % [expon-min, expon-max, v])
+  for exp in to-int(expon-min) to to-int(expon-max + 1.0) do :
+    for std-val in union-std-vals(tol) as List<Double> do :
+      ; println("in-range?(%_, %_)" % [v, c-r(std-val, exp)])
+      if in-range?(v, c-r(std-val, exp)) :
+        val diff = abs(c-r(std-val, exp) - center-value(v))
+        if diff < match-diff :
+          ; std-val component (at the appropriate exp range) is closest to center()
+          match-diff = diff
+          match-val = std-val
+          match-exp = to-double(exp)
+  match-val * pow(10.0, match-exp)
```

Comment:
Here is a proposal
```
public defn between-std-val (v:Toleranced, tol:Double) -> Double :
  val DEFAULT-VALUE = [-1.0, 0.0]
  val expon-min = floor(log10(min-value(v)))
  val expon-max = floor(log10(max-value(v)))

  defn valid? (args: [Double, Int]) -> True|False :
    in-range?(v, make-value(args))

  defn distance-to-center (args: [Double, Int]) -> Double :
    abs(make-value(args) - center-value(v))

  defn make-value (args: [Double, Int]) -> Double :
    val [std-val, exp] = args
    c-r(std-val, exp)

  defn c-r (v:Double e:Int) -> Double :
    v * pow(10.0, to-double(e))

  val exp-range = to-int(expon-min) through to-int(expon-max)

  val valid-pairs = to-tuple $ filter(valid?, zip(union-std-vals(tol), exp-range))

  if empty?(valid-pairs) :
    val closest-std-value = closest-std-val(typ(v), tol)
    println("No standard value matching request v=%_, tol=%_. Using closest std value." % [v, tol, closest-std-value])
    closest-std-value
  else :
    make-value $ minimum(distance-to-center, valid-pairs)
```
`closest-std-val` takes 2 Double, it is possible I am swapping the arguments

Reply:
Accepted your proposal. Looks good in my testing (limited).